uniffi/src/lib.rs

```diff
@@ -23,212 +45,291 @@ pub mod deps {
-    fn lower_into<B: BufMut>(&self, buf: &mut B) {
-        (*self).lower_into(buf);
-    }
-}
-
-impl<T: Lowerable> Lowerable for Option<T> {
-    fn lower_into<B: BufMut>(&self, buf: &mut B) {
-        match self {
-            None => buf.put_u8(0),
-            Some(v) => {
-                buf.put_u8(1);
-                v.lower_into(buf);
-            }
-        }
-    }
-}
-
-impl<T: Lowerable> Lowerable for Vec<T> {
-    fn lower_into<B: BufMut>(&self, buf: &mut B) {
-        let len = u32::try_from(self.len()).unwrap();
-        buf.put_u32(len); // We limit arrays to u32::MAX bytes
-        for item in self.iter() {
-            item.lower_into(buf);
-        }
+/// A helper function to lift a type by deserializing it from a bytebuffer.
+///
+/// For complex types were it's too fiddly or too unsafe to convert them into a special-purpose
+/// C-compatible value, you can use this helper function to implement `lift()` in terms of `read()`
+/// and receive the value as a serialzied byte buffer.
+pub fn try_lift_from_bytebuffer<T: ViaFfi>(buf: ByteBuffer) -> Result<T> {
+    let vec = buf.destroy_into_vec();
+    let mut buf = vec.as_slice();
+    let value = <T as ViaFfi>::try_read(&mut buf)?;
+    if buf.remaining() != 0 {
+        bail!("junk data left in buffer after lifting")
     }
+    Ok(value)
 }
 
-/// Any type that can be received over the FFI must implement the `Liftable` trait, to define how
-/// it gets lifted from bytes into a useable Rust value. We provide default implementtions for
-/// primitive types, and a typical implementation for composite types would lift each member in turn.
-
+/// A helper function to ensure we don't read past the end of a buffer.
+///
+/// Rust won't actually let us read past the end of a buffer, but the `Buf` trait does not support
+/// returning an explicit error in this case, and will instead panic. This is a look-before-you-leap
+/// helper function to instead return an explicit error, to help with debugging.
 pub fn check_remaining<B: Buf>(buf: &B, num_bytes: usize) -> Result<()> {
     if buf.remaining() < num_bytes {
         bail!("not enough bytes remaining in buffer");
     }
     Ok(())
 }
 
-pub trait Liftable: Sized {
-    fn try_lift_from<B: Buf>(buf: &mut B) -> Result<Self>;
-}
-
-pub fn try_lift<T: Liftable>(buf: ByteBuffer) -> Result<T> {
-    let vec = buf.destroy_into_vec();
-    let mut buf = vec.as_slice();
-    let value = <T as Liftable>::try_lift_from(&mut buf)?;
-    if buf.remaining() != 0 {
-        bail!("junk data left in buffer after deserializing")
-    }
-    Ok(value)
-}
-
-macro_rules! impl_lowerable_liftable_for_num_primitive {
+/// Blanket implementation of ViaFfi for numeric primitives.
+///
+/// Numeric primitives have a straightforward mapping into C-compatible numeric types,
+/// sice they are themselves a C-compatible numeric type!
+macro_rules! impl_via_ffi_for_num_primitive {
     ($($T:ty,)+) => { impl_lowerable_liftable_for_num_primitive!($($T),+); };
```

Comment:
```suggestion
    ($($T:ty,)+) => { impl_via_ffi_for_num_primitive!($($T),+); };
```

Reply:
I am lightly alarmed that this did not cause any test failures...